superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,459 @@
+    object: t.Optional[transformers.Trainer] = None
+    model_kwargs: t.Dict = dc.field(default_factory=dict)
+    tokenizer_kwags: t.Dict = dc.field(default_factory=dict)
+    prompt_template: str = "{input}"
+    prompt_func: t.Optional[t.Callable] = dc.field(default=None)
+
+    def __post_init__(self):
+        self.identifier = self.identifier or self.model_name_or_path
+        # overwrite model kwargs
+        if self.bits is not None:
+            self.model_kwargs["load_in_4bit"] = self.bits == 4
+            self.model_kwargs["load_in_8bit"] = self.bits == 8
+        super().__post_init__()
+
+    def init_model_and_tokenizer(self):
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_name_or_path,
+            **self.model_kwargs,
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name_or_path,
+            **self.tokenizer_kwags,
+        )
+        tokenizer.pad_token = tokenizer.pad_token or tokenizer.unk_token
+        return model, tokenizer
+
+    def create_trainer(
+        self, train_dataset, eval_dataset, training_args, **kwargs
+    ) -> transformers.Trainer:
+        trainer = transformers.Trainer(
+            model=self.model,
+            tokenizer=self.tokenizer,
+            args=training_args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            **kwargs,
+        )
+        return trainer
+
+    def init(self):
+        self.prompter = Prompter(self.prompt_template, self.prompt_func)
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+    def _fit(
+        self,
+        X: t.Any,
+        y: t.Optional[t.Any] = None,
+        configuration: t.Optional[_TrainingConfiguration] = None,
+        data_prefetch: bool = False,
+        db: t.Optional["Datalayer"] = None,
+        metrics: t.Optional[t.Sequence["Metric"]] = None,
+        select: t.Optional["Select"] = None,
+        validation_sets: t.Optional[t.Sequence[t.Union[str, "_Dataset"]]] = None,
+        **kwargs,
+    ):
+        assert configuration is not None, "configuration must be provided"
+
+        training_args = LLMTrainingArguments(**configuration.kwargs)  # type: ignore
+
+        # get device map
+        device_map: t.Union[None, str, t.Dict[str, int]] = None
+        if os.environ.get("LOCAL_RANK") is not None:
+            device_map = {"": int(os.environ.get("LOCAL_RANK", "0"))}
+        elif torch.backends.mps.is_available():
+            device_map = "mps"
+
+        quantization_config = self._create_quantization_config(training_args)
+
+        self.model_kwargs["quantization_config"] = quantization_config
+        self.model_kwargs["device_map"] = device_map
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+        self.tokenizer.model_max_length = (
+            training_args.max_length or self.tokenizer.model_max_length
+        )
+        self._prepare_lora_training(training_args)
```

Comment:
@jieguangzhou 
In the future, 
we should have strategies classes
like LoraTrainer, PeftTrainer, QLoraTrainer, etc

which could be used to train

cc @blythed

Reply:
I agree, also something which the user can easily customize.